ACAP-1017: add AEM Assets script/plugin for image slots

iivvaannxx opened the pull request:
## Context

This PR enables a set of functions/utilities that will be used in subsequent PRs to support AEM Asset images in the different dropins slots, and URL construction for those that currently don't have them, namely PLP, PREX and LiveSearch.

> [!NOTE]
> I wanted to provide a type safe experience, so I wrote a bunch of JSDoc `typedef` declarations at the bottom of the file. I would have liked a `.d.ts` better but that's out of the scope of this PR.

## Jira Ticket

[ACAP-1017](https://jira.corp.adobe.com/browse/ACAP-1017)

## Requirements

@sirugh To add the `commerce-assets-enabled` in the default starter content sheet configs.

## Example Usages

### For Slots
For slots, a `width` is required, and the `size` optimization parameter is dropped to avoid confusion (see L154).

```js
// Other imports ...
import { tryRenderAemAssetsImage } from '../../scripts/assets.js';

// Some dropin container...
slots: {
  Thumbnail: async (ctx) => {
    await tryRenderAemAssetsImage(ctx, {
      alias: 'my-seo-name',
      params: {
        width: 200,
        // other params...
      },
      src: "raw image src",
      // wrapper: optional wrapper -> document.createElement('a')
    });
  },
  }
//
```

### URL Construction
URL construction allows to use any parameters you want, given an `url` and an `alias` and you get the AEM Assets compatible url back.

```js
// Other imports ...
import { generateAemAssetsOptimizedUrl } from '../../scripts/assets.js';

// Some other code...
const aemAssetsOptimizedUrl = generateAemAssetsOptimizedUrl(aemAssetsImageSrc, 'my-seo-name', {
  format: 'png',
  flip: 'v',
  rotate: 180,
  size: {
    width: 200,
    height: 600
  }
});
```

iivvaannxx commented:
Last commit adds a method that conditionally tries to generate a url for AEM Assets similarly to `tryRenderAemAssetsImage`, intended to be used by non-dropin blocks like PREX, PLP and LiveSearch, so they don't have to repeat the same condition all over the place

sirugh reviewed (approved):
We can merge this now, and whenever dropins expose requisite slots they can add the asset integration code.

For now we just document the new commerce-assets-enabled flag.

herzog31 commented:
I would prefer having the type definitions in a place that is not loaded by the website visitor :)

iivvaannxx commented:
@herzog31 done!

sirugh commented on scripts/assets.js:
```diff
@@ -0,0 +1,129 @@
+import { getConfigValue } from './configs.js';
+
+/** Reports whether AEM Assets are being used as the asset source. */
+export async function isAemAssetsEnabled() {
+  const config = await getConfigValue('commerce-assets-enabled');
+  return config === 'true';
+}
```
FYI this will be changing soon and will no longer require `await` :)

https://github.com/hlxsites/aem-boilerplate-commerce/pull/387/files#diff-eb55b6bcdfbd2110ceed72aa446694d87208d331fbcf59d2887701c142611ccfR201

iivvaannxx commented:
Changed `base` branch to `integration` and rebased PR. 

@sirugh Cypress tests seem to take quite a bit longer and some are failing, is this normal? The ones that have this problem should not be related to any change (at the time of writing) of this PR at all.

sirugh commented:
> @sirugh Cypress tests seem to take quite a bit longer and some are failing, is this normal? The ones that have this problem should not be related to any change (at the time of writing) of this PR at all.

Always check `main` - there are failures there, right now. Dev is aware, and we will investigate tomorrow, unless you have time and want to try and fix them :D

iivvaannxx commented:
> Always check `main` - there are failures there, right now. Dev is aware, and we will investigate tomorrow, unless you have time and want to try and fix them :D

Didn't check, will do 🫡. Don't have time to check the issue though 🫠
